crates/telio-nurse/src/aggregator.rs

```diff
@@ -0,0 +1,808 @@
+    ) {
+        self.change_peer_state_common(
+            event,
+            PeerEndpointTypes {
+                initiator_ep: initiator_ep.into(),
+                responder_ep: responder_ep.into(),
+            },
+        )
+    }
+
+    #[allow(dead_code)]
+    pub fn change_peer_state_relayed(&mut self, event: AnalyticsEvent) {
+        self.change_peer_state_common(
+            event,
+            PeerEndpointTypes {
+                initiator_ep: EndpointType::Relay,
+                responder_ep: EndpointType::Relay,
+            },
+        )
+    }
+
+    fn change_peer_state_common(&mut self, event: AnalyticsEvent, endpoints: PeerEndpointTypes) {
+        if !self.aggregate_nat_traversal_events {
+            return;
+        }
+
+        let mut data_guard = self.data.write();
+        let event_entry = data_guard.current_peer_events.entry(event.public_key);
+        let new_segment = match event_entry.borrow() {
+            Entry::Occupied(entry)
+                if entry.get().0.peer_state != event.peer_state || entry.get().1 != endpoints =>
+            {
+                Some(ConnectionSegmentData::new_peer(
+                    &entry.get().0,
+                    entry.get().1,
+                    event.timestamp.into(),
+                    event.rx_bytes,
+                    event.tx_bytes,
+                ))
+            }
+            _ => None,
+        };
+
+        if event.peer_state != PeerState::Connected {
+            if let Entry::Occupied(entry) = event_entry {
+                entry.remove();
+            }
+        } else if let Entry::Vacant(entry) = event_entry {
+            entry.insert((event, endpoints));
+        }
+
+        data_guard.unacknowledged_segments.extend(new_segment);
+    }
+
+    #[allow(dead_code)]
+    pub fn change_relay_state(
+        &mut self,
+        event: AnalyticsEvent,
+        reason: RelayConnectionChangeReason,
+    ) {
+        if !self.aggregate_relay_events {
+            return;
+        }
+
+        let mut data_guard = self.data.write();
+
+        let new_segment = data_guard
+            .current_relay_event
+            .as_ref()
+            .filter(|old_event| {
+                old_event.public_key != event.public_key || old_event.peer_state != event.peer_state
+            })
+            .map(|old_event| ConnectionSegmentData::new_relay(old_event, event.timestamp.into()));
+
+        if let Some(segment) = new_segment {
+            data_guard.unacknowledged_segments.push(segment);
+        }
+
+        if new_segment.is_some() || data_guard.current_relay_event.is_none() {
+            data_guard.current_relay_event = Some(RelayEvent {
+                public_key: event.public_key,
+                peer_state: event.peer_state,
+                timestamp: event.timestamp.into(),
+                reason,
+            })
+        }
+    }
+
+    #[allow(dead_code)]
+    pub async fn collect_unacknowledged_segments(&self) -> Vec<ConnectionSegmentData> {
```

Comment:
How is the `ConnectivityDataAggregator` supposed to be used by libtelio? The way I understand this design, there will be one instance of it to which we will push events (both nat and relay) when something changes. At heartbeat time we will call `collect_unacknowledged_segments` and send the result via the 2 new fields `NAT_monitoring` and `derp_monitoring`. If my understanding is correct I see one issue. There are 2 separate fields but we get one collection from `collect_unacknowledged_segments`. Since I don't think we have any use for such a collection, can we return tuple with separate collections for nat and relay events?

Also, I think it would be simlper to debuging to sort those collections based on timestamp before returning them.